pkg/controller/resourcequota/resource_quota_controller.go

```diff
@@ -423,24 +423,90 @@ func (rq *ResourceQuotaController) Sync(discoveryFunc NamespacedResourcesFunc, p
 			return
 		}
 
-		// Something has changed, so track the new state and perform a sync.
-		klog.V(2).Infof("syncing resource quota controller with updated resources from discovery: %v", newResources)
-		oldResources = newResources
-
 		// Ensure workers are paused to avoid processing events before informers
 		// have resynced.
 		rq.workerLock.Lock()
 		defer rq.workerLock.Unlock()
 
-		// Perform the monitor resync and wait for controllers to report cache sync.
-		if err := rq.resyncMonitors(newResources); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
-			return
+		// Once we get here, we should not unpause workers until we've successfully synced
+		attempt := 0
+		wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
+			attempt++
+
+			// On a reattempt, check if available resources have changed
+			if attempt > 1 {
+				// Get the current resource list from discovery.
+				newResources, err = GetQuotableResources(discoveryFunc)
+				if err != nil {
+					utilruntime.HandleError(err)
+					if discovery.IsGroupDiscoveryFailedError(err) && len(newResources) > 0 {
+						// In partial discovery cases, don't remove any existing informers, just add new ones
+						for k, v := range oldResources {
+							newResources[k] = v
```

Comment:
Is this correct? This line seems to overwrite values but the comment seems to imply we only want to add values if the keys are non-existing.

Reply:
Sorry, please ignore, I realized later that this was basically just a set thing.